SAP/spartacus
# feat: new `FeatureToggles` config to drive parts of Spartacus config, e.g. cms mapping

The new `FeatureToggles` config is separate from Spartacus config, to be able to control the global Spartacus config. 
The `FeatureToggles` is a simple list of boolean properties which allows for enabling breaking changes in Spartacus, ALSO changes in the default config chunks like cms mapping etc. See example:
```ts
provideDefaultFeatureToggles({
  // boolean is a simple, serializable flag, so can be even potentially
  // fetched from remote configuration endpoint in runtime
  myAccountV2: true
})
```
```ts
// conditionally changing the default config chunk based on the simple boolean feature toggle:
provideConfigFactory(()=>{
  const featureToggles = inject(FeatureToggles);
  return featureToggles.myAccountV2
    ? { cmsComponents: { component: MyAccountComponentV2 } }
    : { cmsComponents: { component: MyAccountComponent } };
})
```

In other words, the `FeatureToggles` config can be a bridge between a simple (and even remote!) feature toggles configuration of Spartacu and making actual changes into the default config chunks in Spartacus (like default cms mapping).

Note: In this PR added also an ESLint rule to prefer using in our libraries the function `provideDefaultFeatureToggles()` over `provideFeatureToggles()` - so our providers have lower precedence than customers' ones.

**ALL FEATURE TOGGLES DEFINED IN THE CORE LIB:**

All new toggles are now added in the file `feature-toggles.ts` for customers and in `spartacus-features.module.ts` for our storefront app.

The centralization of feature toggles was done because:
- for doc writers: otherwise It was hard to track down new toggles each release due to distribution across many libs
- for devs: adding a new toggle lacked a standard process, raising the question: Where to define it? This dilemma leads to issues, **especially when default toggles values were provided by mistake in lazy-loaded modules it was not working** (because the feature toggles mechanism relies on the **root injector**).
- this confusion wasted developers' time, as they "just want to add a toggle" but get stuck figuring out where to place it, with the process varying across different libraries (storefront, core, feature).


**DEPRECATIONS:**

This PR deprecates the `FeaturesConfig` interface in favor of `FeatureToggles`: So please use:
```ts
provideFeatureToggles({
  [string]: boolean
})
```
instead of
```ts
provideConfig({
  features: {
    [string]: boolean
  }
})
```
That said, the old approach will still work until we remove the `FeaturesConfig` eventually.

**QA steps**

I) verify that FeatureToggles mechanism work as expected on the example property `a11yRequiredAsterisks`:
1. open login page
2. verify that the asterisk regarding required form fields IS displayed
3. in `spartacus-features.module.ts` REMOVE the toggle `a11yRequiredAsterisks` and verify asterisk is NOT displayed
4. in `default-feature-toggles.ts` set `a11yRequiredAsterisks: true` and verify asterisk IS displayed
5. in `spartacus-features.module.ts` set `a11yRequiredAsterisks: false` and verify asterisk is NOT displayed
6. in `spartacus-features.module.ts` set `a11yRequiredAsterisks: true` and verify asterisk is IS displayed
7. in `default-feature-toggles.ts` set `a11yRequiredAsterisks: false` and verify asterisk IS STILL displayed
8. in `spartacus-features.module.ts` REMOVE the toggle `a11yRequiredAsterisks` and verify asterisk is NOT displayed

II) verify FeatureToggles mechanism can alter a cms mapping, let's use `a11yRequiredAsterisks` for the spike purposes:
1. in `login-form.module.ts` add to the `providers` array the following code:
```ts
    provideDefaultConfigFactory(() => {
      return inject(FeatureToggles).a11yRequiredAsterisks
        ? {
            cmsComponents: {
              ReturningCustomerLoginComponent: {
                component: SpikeComponent,
              },
            },
          }
        : {};
    }),
```
2. ...and in the top of the same file (`login-form.module.ts`) add the following code:
```ts
@Component({
  selector: 'cx-spike',
  template: 'SPIKE COMPONENT WITH ASTERISK (*.*)',
})
export class SpikeComponent {}
```

3. Visit the login page and verify that the `SPIKE` text is displayed
4. in `spartacus-features.module.ts` REMOVE the toggle `a11yRequiredAsterisks` and verify that the `SPIKE COMPONENT WITH ASTERISK (*.*)` text is NOT displayed

III) verify the old FeaturesConfig mechanism cannot alter cms mapping, but still works for enabling features:
1. ensure that `a11yRequiredAsterisks` in `provideFeatureToggles` is set to `false`
2. in `spartacus-features.module.ts` add to the `providers` array the following code `provideConfig({features: { a11yRequiredAsterisks: true  }})` and verify that the ORIGINAL LOGIN COMPONENT is displayed WITH real a11y asterisk feature displayed
3. in `spartacus-features.module.ts` set `a11yRequiredAsterisks: false` inside the above mentioned `provideConfig()` and verify that the ORIGINAL LOGIN COMPONENT is displayed WITHOUT real a11y asterisk feature displayed
4. in `spartacus-features.module.ts` set `a11yRequiredAsterisks: true` inside the `provideFeatureToggles()` and verify that the the `SPIKE` text is displayed even if `a11yRequiredAsterisks: false` inside the `provideConfig()` is false


closes https://jira.tools.sap/browse/CXSPA-6586
